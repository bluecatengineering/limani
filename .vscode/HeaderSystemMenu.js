/*
Copyright 2023 BlueCat Networks Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import { t } from '@bluecateng/l10n.macro';
import {
    HeaderIcon,
    Layer,
    Menu,
    MenuItem,
    useMenuHandler,
} from '@bluecateng/pelagos';
import { Settings } from '@carbon/icons-react';
import PropTypes from 'prop-types';
import usePageMessages from '../components/usePageMessages';
import usePlatformData from '../hooks/usePlatformData';
import { resizeButtonMenu } from '../utils/display';
import './HeaderSystemMenu.less';

/**
 * HeaderSystemMenu component is a button when clicked this button reveals the
 * links to System menu options. <br>
 * With user access permission, two options for retrieving logs are available.
 * "Download all logs" retrieves all logs as a single file, and "View logs"
 * shows the last 1000 lines of relevant Gateway logs, providing searching and
 * filtering options. <br/>
 * This component is intended to be nested inside the PlatformDataContext as
 * it will require access to PlatformData. <br>
 * The component should always be wrapped inside parent element that
 * has an id=root, all the pages in Gateway is wrapped with a top-level parent
 * element that has an id=root.
 */

const HeaderSystemMenu = ({ className }) => {
    const { addMessages } = usePageMessages();
    const { data } = usePlatformData();

    const canDownloadLogs = data?.user?.permissions.download_logs;
    const canViewLogs = data?.user?.permissions.view_logs;
    const downloadLogs = () => {
        // doGet('/admin/logs/download')
        fetch('http://localhost:5000/admin/logs/download')
            .then((resp) => {
                if (!resp.ok) {
                    addMessages([{ 'type': 'error', 'text': resp.message }]);
                }
                console.log({ resp });
                const check = resp.headers;
                const check2 = resp.headers.get('content-disposition');
                console.log({ check, check2 });
                return new Blob([resp.data]);
            })
            .then((blob) => {
                console.log({ blob });
                const url = window.URL.createObjectURL(new Blob([blob]));
                const link = document.createElement('a');
                link.href = url;
                let name = new Date().toISOString().slice(0, 19);
                name = name.replaceAll(':', '-').replace('T', '_');
                link.setAttribute('download', `logs_${name}.zip`);
                link.style.display = 'none'; // Hide the anchor element
                document.body.appendChild(link);
                link.click();
            })
            .catch((error) => {
                console.log({ error });
                if (error?.status === 404) {
                    console.log(404);
                    addMessages([{ 'type': 'warning', 'text': error.message }]);
                } else {
                    console.log('other');
                    addMessages([{ 'type': 'error', 'text': error.message }]);
                }
            });
    };

    const viewLogs = () => window.location.replace('/admin/view_logs');

    const { expanded, buttonProps, menuProps, guardProps } = useMenuHandler();

    resizeButtonMenu(expanded, 'systemMenuButton', 'systemMenu');

    return (
        <>
            {(canDownloadLogs || canViewLogs) && (
                <div className={className}>
                    <button
                        id='systemMenuButton'
                        className='HeaderSystemMenu__button'
                        type='button'
                        aria-controls={expanded ? 'systemMenu' : null}
                        aria-haspopup='true'
                        aria-expanded={expanded}
                        {...buttonProps}>
                        <HeaderIcon
                            className='button'
                            icon={Settings}
                            label={t`System`}
                        />
                    </button>
                    {expanded && (
                        <Layer
                            id='systemMenu'
                            role='menu'
                            className='HeaderSystemMenu__menu'
                            aria-label={`System Menu`}>
                            <div tabIndex={0} {...guardProps} />
                            <Menu {...menuProps}>
                                {canDownloadLogs && (
                                    <MenuItem
                                        id='systemMenu-dl_logs'
                                        onClick={
                                            downloadLogs
                                        }>{t`Download all logs`}</MenuItem>
                                )}
                                {canViewLogs && (
                                    <MenuItem
                                        id='systemMenu-vw_logs'
                                        onClick={
                                            viewLogs
                                        }>{t`View logs`}</MenuItem>
                                )}
                            </Menu>
                            <div tabIndex={0} {...guardProps} />
                        </Layer>
                    )}
                </div>
            )}
        </>
    );
};

HeaderSystemMenu.propTypes = {
    /** The component class name(s). */
    className: PropTypes.string,
};

export default HeaderSystemMenu;
