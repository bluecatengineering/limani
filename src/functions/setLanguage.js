/*
Copyright 2023 BlueCat Networks Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import l10n from '../l10n';

const loaders = {
    en: () => import('../l10n/en.po'),
    zz: () => import('../l10n/zz.po'),
};

/**
 * Sets the language for Limani. The default language is English, there is no
 * need to call this function unless the selected language is different.
 * This function should be called as early as possible,
 * the caller must catch the returned promise and handle any errors.
 * @param {'en'|'es'} language The language translated content to load for.
 * @returns {*|Promise<never>}
 *
 * @example
 * import {setLanguage} from '@bluecateng/limani';
 *
 * setLanguage('es').catch((error) => {
 *     // handle error
 * })
 */
const setLanguage = (language) => {
    const loader = loaders[language];
    return loader
        ? loader().then(({ default: data }) => l10n.load(data))
        : Promise.resolve(console.warn(`Unknown language ${language}`));
};

export default setLanguage;
