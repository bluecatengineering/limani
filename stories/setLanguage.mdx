{/* Copyright 2023 BlueCat Networks Inc */}
{/*  */}
{/* Permission is hereby granted, free of charge, to any person obtaining a copy */}
{/* of this software and associated documentation files (the "Software"), to deal */}
{/* in the Software without restriction, including without limitation the rights */}
{/* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell */}
{/* copies of the Software, and to permit persons to whom the Software is */}
{/* furnished to do so, subject to the following conditions: */}
{/*  */}
{/* The above copyright notice and this permission notice shall be included in all */}
{/* copies or substantial portions of the Software. */}
{/*  */}
{/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR */}
{/* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, */}
{/* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE */}
{/* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER */}
{/* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, */}
{/* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE */}
{/* SOFTWARE. */}
import { Meta } from '@storybook/addon-docs';

<Meta title='functions/setLanguage' />

# setLanguage

Sets the language for Limani. The default language is set to English i.e. `en`.
There is no need to call this function unless the selected
language is other than `en`. This function should be called as early as
possible, and wait before rendering the components,as it's an asynchronous operation.
Currently only `en` is supported.

## Parameters

<table>
    <thead>
        <tr>
            <th>Param</th>
            <th>Type</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>language</td>
            <td>`String`</td>
            <td>The language that needs to be set.</td>
        </tr>
    </tbody>
</table>

## Returns

`Promise` - Resolves successfully on language change.

## Example

```js
import { setLanguage } from '@bluecateng/limani';

const Example = () => (
    setLanguage('es').catch((error) => {
        // handle error
    })
    ...
);
```
