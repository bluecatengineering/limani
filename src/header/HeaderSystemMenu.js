/*
Copyright 2023 BlueCat Networks Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import PropTypes from 'prop-types';
import { faCog } from '@fortawesome/free-solid-svg-icons';
import {
    HeaderIcon,
    useMenuHandler,
    Layer,
    Menu,
    MenuItem,
} from '@bluecateng/pelagos';
import usePlatformData from '../hooks/usePlatformData';
import { resizeButtonMenu } from '../utils/display';
import { t } from '@bluecateng/l10n.macro';
import './HeaderSystemMenu.less';

/**
 * HeaderSystemMenu component is a button when clicked this button reveals the
 * links to System menu options. <br>
 * With user access permission, Download logs is an available option to retrieve
 * all logs as a single file. <br>
 * This component is intended to be nested inside the PlatformDataContext as
 * it will require access to PlatformData. <br>
 * The component should always be wrapped inside parent element that
 * has an id=root, all the pages in Gateway is wrapped with a top-level parent
 * element that has an id=root.
 */

const HeaderSystemMenu = ({ className }) => {
    const { data } = usePlatformData();
    const canDownloadLogs = data?.user?.permissions.download_logs;
    const canViewLogs = data?.user?.permissions.view_logs;
    const downloadLogs = () =>
        window.open('/admin/logs/download', '_blank').focus();
    const viewLogs = () => window.open('/admin/view_logs', '_blank').focus();

    const { expanded, buttonProps, menuProps, guardProps } = useMenuHandler();

    resizeButtonMenu(expanded, 'systemMenuButton', 'systemMenu');

    return (
        <>
            {(canDownloadLogs || canViewLogs) && (
                <div className={className}>
                    <button
                        id='systemMenuButton'
                        className='HeaderSystemMenu__button'
                        type='button'
                        aria-controls={expanded ? 'systemMenu' : null}
                        aria-haspopup='true'
                        aria-expanded={expanded}
                        {...buttonProps}>
                        <HeaderIcon
                            className='button'
                            icon={faCog}
                            label={t`System`}
                        />
                    </button>
                    {expanded && (
                        <Layer
                            id='systemMenu'
                            role='menu'
                            className='HeaderSystemMenu__menu'
                            aria-label={`System Menu`}>
                            <div tabIndex={0} {...guardProps} />
                            <Menu {...menuProps}>
                                {canDownloadLogs && (
                                    <MenuItem
                                        id='systemMenu-dl_logs'
                                        onClick={
                                            downloadLogs
                                        }>{t`Download logs`}</MenuItem>
                                )}
                                {canViewLogs && (
                                    <MenuItem
                                        id='systemMenu-vw_logs'
                                        onClick={
                                            viewLogs
                                        }>{t`View logs`}</MenuItem>
                                )}
                            </Menu>
                            <div tabIndex={0} {...guardProps} />
                        </Layer>
                    )}
                </div>
            )}
        </>
    );
};

HeaderSystemMenu.propTypes = {
    /** The component class name(s). */
    className: PropTypes.string,
};

export default HeaderSystemMenu;
