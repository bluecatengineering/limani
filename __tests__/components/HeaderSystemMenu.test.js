/*
Copyright 2023 BlueCat Networks Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import { shallow } from 'enzyme';
import HeaderSystemMenu from '../../src/header/HeaderSystemMenu';
import setLanguage from '../../src/functions/setLanguage';
import { useMenuHandler } from '@bluecateng/pelagos';

jest.unmock('../../src/header/HeaderSystemMenu');

jest.mock('../../src/hooks/usePlatformData', () =>
    jest
        .fn()
        .mockReturnValueOnce({
            data: {
                user: {
                    permissions: { download_logs: false, view_logs: false },
                },
            },
        })
        .mockReturnValueOnce({
            data: {
                user: { permissions: { download_logs: true, view_logs: true } },
            },
        })
        .mockReturnValueOnce({
            data: {
                user: { permissions: { download_logs: true, view_logs: true } },
            },
        })
        .mockReturnValueOnce({
            data: {
                user: { permissions: { download_logs: true, view_logs: true } },
            },
        })
        .mockReturnValueOnce({
            data: {
                user: { permissions: { download_logs: true, view_logs: true } },
            },
        }),
);

jest.mock('@bluecateng/pelagos');

describe('HeaderSystemMenu', () => {
    describe('Rendering', () => {
        it('Render HeaderSystemMenu component with default props', () => {
            useMenuHandler.mockReturnValueOnce({
                expanded: true,
                buttonProps: null,
                menuProps: null,
                guardProps: null,
            });
            const wrapper = shallow(<HeaderSystemMenu />);
            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('Render HeaderSystemMenu component with props', () => {
            useMenuHandler.mockReturnValueOnce({
                expanded: false,
                buttonProps: { onClick: () => {}, onKeyDown: () => {} },
                menuProps: null,
                guardProps: null,
            });
            const wrapper = shallow(
                <HeaderSystemMenu className='customSystemMenu' />,
            );
            expect(wrapper.getElement()).toMatchSnapshot();
        });
        it('Render HeaderSystemMenu component with pseudo translated text', () => {
            useMenuHandler.mockReturnValueOnce({
                expanded: false,
                buttonProps: { onClick: () => {}, onKeyDown: () => {} },
                menuProps: null,
                guardProps: null,
            });
            setLanguage('zz').finally(() => {
                const wrapper = shallow(
                    <HeaderSystemMenu className='customSystemMenu' />,
                );
                expect(wrapper.getElement()).toMatchSnapshot();
            });
        });
        it('Test the view logs click functionality', () => {
            // Mock the current window location
            const mockWindow = jest.fn();
            global.window.location = { replace: mockWindow };

            // Mock the system menu button's menu handler
            useMenuHandler.mockReturnValueOnce({
                expanded: true,
                buttonProps: null,
                menuProps: null,
                guardProps: null,
            });
            // Render the HeaderSystemMenu element
            const wrapper = shallow(
                <HeaderSystemMenu className='customSystemMenu' />,
            );

            // Click the view logs menu option
            wrapper.find('#systemMenu-vw_logs').simulate('click');

            // Check redirect to view_logs URL
            expect(mockWindow).toHaveBeenCalledWith('/admin/view_logs');
        });
        it('Test the download all logs click functionality', () => {
            // Mock the opening of a new window
            const mockWindowOpen = jest
                .fn()
                .mockReturnValue({ focus: jest.fn() });
            global.window.open = mockWindowOpen;

            // Mock the system menu button's menu handler
            useMenuHandler.mockReturnValueOnce({
                expanded: true,
                buttonProps: null,
                menuProps: null,
                guardProps: null,
            });
            // Render the HeaderSystemMenu element
            const wrapper = shallow(
                <HeaderSystemMenu className='customSystemMenu' />,
            );
            // Click the download all logs button
            wrapper.find('#systemMenu-dl_logs').simulate('click');

            // Check a new window was opened and redirected to download logs URL
            expect(mockWindowOpen).toHaveBeenCalledWith(
                '/admin/logs/download',
                '_blank',
            );
            // Check download all logs window was put into focus
            expect(mockWindowOpen().focus).toHaveBeenCalled();
        });
    });
});
