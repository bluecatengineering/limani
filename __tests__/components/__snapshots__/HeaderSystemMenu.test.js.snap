// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderSystemMenu Rendering Render HeaderSystemMenu component with default props 1`] = `<React.Fragment />`;

exports[`HeaderSystemMenu Rendering Render HeaderSystemMenu component with props 1`] = `
<React.Fragment>
  <div
    className="HeaderSystemMenu customSystemMenu"
  >
    <button
      aria-controls={null}
      aria-expanded={false}
      aria-haspopup="true"
      className="HeaderSystemMenu__button"
      id="systemMenuButton"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      <AppShellIcon
        icon={[Function]}
      />
    </button>
  </div>
</React.Fragment>
`;

exports[`HeaderSystemMenu Rendering Render HeaderSystemMenu component with pseudo translated text 1`] = `
<React.Fragment>
  <div
    className="HeaderSystemMenu customSystemMenu"
  >
    <button
      aria-controls={null}
      aria-expanded={false}
      aria-haspopup="true"
      className="HeaderSystemMenu__button"
      id="systemMenuButton"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      <AppShellIcon
        icon={[Function]}
      />
    </button>
  </div>
</React.Fragment>
`;
